### PR TITLE
BUG: Fix bug when using df.query with "str.contains()"

### PR DESCRIPTION
**Pull request by dtpc**
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Fixes bug in on line 588 of `pandas/core/computation/expr.py`: `AttributeError: 'dict' object has no attribute 'append'`

Also addresses the specific example in #22172

Example:
```python
import pandas as pd
df = pd.DataFrame([["I", "XYZ"], ["IJ", None]], columns=['A', 'B'])
df.query("B.str.contains('Z', na=False)", engine="python")
```

```
Traceback (most recent call last):
  File "./test_query.py", line 12, in <module>
    result = df.query("B.str.contains('Z', na=False)", engine="python")
  File "./lib/python3.6/site-packages/pandas/core/frame.py", line 2847, in query
    res = self.eval(expr, **kwargs)
  File "./lib/python3.6/site-packages/pandas/core/frame.py", line 2962, in eval
    return _eval(expr, inplace=inplace, **kwargs)
  File "./lib/python3.6/site-packages/pandas/core/computation/eval.py", line 291, in eval
    truediv=truediv)
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 739, in __init__
    self.terms = self.parse()
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 756, in parse
    return self._visitor.visit(self.expr)
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 321, in visit
    return visitor(node, **kwargs)
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 327, in visit_Module
    return self.visit(expr, **kwargs)
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 321, in visit
    return visitor(node, **kwargs)
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 330, in visit_Expr
    return self.visit(node.value, **kwargs)
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 321, in visit
    return visitor(node, **kwargs)
  File "./lib/python3.6/site-packages/pandas/core/computation/expr.py", line 588, in visit_Call_35
    kwargs.append(ast.keyword(
AttributeError: 'dict' object has no attribute 'append'
```

### CODE REVIEW

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you add a whatsnew note for v0.25?

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -322,7 +322,7 @@ Sparse
 Other
 ^^^^^
 
--
+- Bug in :class:`BaseExprVisitor` for using named kwargs in :func:`eval` expressions.
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.25.0.rst]: Can you come up with a more user-facing message? 

Also if the focus is on passing keywords args through eval can you add a test which explicitly does this?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/computation/test_eval.py [hunk 2]
```diff
@@ -1903,3 +1903,15 @@ def test_validate_bool_args(self):
         for value in invalid_values:
             with pytest.raises(ValueError):
                 pd.eval("2+2", inplace=value)
+
+
+def test_query_str_contains():
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: add the issue number to this test. 

also name the test something more constructive. maybe test_with_str_accessor

see if this test can fit elsewhere in this file in a more logical section.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
@dtpc can you merge master and address comments?

**Comment by jreback** [comment]
can you merge master and update

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (dtpc) [reply doc/source/whatsnew/v0.25.0.rst]: I've updates the tests and whatsnew message

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (dtpc) [reply pandas/tests/computation/test_eval.py]: Merged and updated tests

File: doc/source/whatsnew/v0.25.0.rst [hunk 3]
```diff
@@ -412,6 +412,7 @@ Other
 
 - Improved :class:`Timestamp` type checking in various datetime functions to prevent exceptions when using a subclassed `datetime` (:issue:`25851`)
 - Bug in :class:`Series` and :class:`DataFrame` repr where ``np.datetime64('NaT')`` and ``np.timedelta64('NaT')`` with ``dtype=object`` would be represented as ``NaN`` (:issue:`25445`)
+- Bug in :class:`BaseExprVisitor` which caused :func:`eval` expressions to fail when named keyword arguments were included within the expression string (:issue:`25813`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.25.0.rst]: Still needs a little work to make this user facing, i.e. only references items in the documented API (which BaseExprVisitor is not).

Would be better referencing the DataFrame.query method

File: pandas/tests/computation/test_eval.py [hunk 4]
```diff
@@ -1757,6 +1757,29 @@ def test_no_new_globals(self, engine, parser):
         assert gbls == gbls2
 
 
+class TestEvalNamedKWargs(object):
```
Comment (WillAyd) [inline pandas/tests/computation/test_eval.py]: I don't think we need a class here; OK to remove this and put tests at top level of module

File: pandas/tests/computation/test_eval.py [hunk 5]
```diff
@@ -1757,6 +1757,29 @@ def test_no_new_globals(self, engine, parser):
...
+        def fn(x):
+            return 2 * x
+        result = pd.eval("fn(x=2)", engine=engine, parser=parser)
+        assert result == 4
+
+    def test_query_str_contains(self, parser):
+        df = pd.DataFrame([["I", "XYZ"], ["IJ", None]], columns=['A', 'B'])
+
+        expected = df[df["A"].str.contains("J")]
+        result = df.query("A.str.contains('J')", engine="python",
```
Comment (WillAyd) [inline pandas/tests/computation/test_eval.py]: Any reason to not parametrize this for all engines?

File: pandas/tests/computation/test_eval.py [hunk 6]
```diff
@@ -1757,6 +1757,29 @@ def test_no_new_globals(self, engine, parser):
         assert gbls == gbls2
 
 
+class TestEvalNamedKWargs(object):
+    # xref https://github.com/pandas-dev/pandas/issues/25813
```
Comment (WillAyd) [inline pandas/tests/computation/test_eval.py]: Just add `#GH 25813` to the first line of the tests added here

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@pandas-dev/pandas-core if someone wants to finish up

File: pandas/tests/computation/test_eval.py (hunk 5 above)
Comment (gfyoung) [reply pandas/tests/computation/test_eval.py]: Yes, because `engine=pandas` fails for this test.  It's tries to hash a `Series` object.

**Comment by jreback** [comment]
this PR is superseded by #26426

**Pull request closed without merging**